tsfc/loopy.py

```diff
@@ -43,41 +43,52 @@ def pym_multiindex(self, multiindex):
                 indices.append(index)
         return tuple(indices)
 
-    def pymbolic_variable(self, node):
+    def gem_to_pym(self, node):
         try:
             pym = self.gem_to_pymbolic[node]
         except KeyError:
             name = self.name_gen(node.name)
             pym = p.Variable(name)
             self.gem_to_pymbolic[node] = pym
+        return pym
+
+    def pymbolic_variable(self, node):
+        pym = self.gem_to_pym(node)
         if node in self.indices:
             indices = self.pym_multiindex(self.indices[node])
             if indices:
                 return p.Subscript(pym, indices)
-            else:
-                return pym
-        else:
-            return pym
+        return pym
+
+    def pymbolic_variable_and_destruct(self, node):
+        pym = self.gem_to_pym(node)
+        if node in self.indices:
+            indices = self.pym_multiindex(self.indices[node])
+            if indices:
+                return (pym, indices)
+        return (pym, ())
 
     def active_inames(self):
         # Return all active indices
         return frozenset([i.name for i in self.active_indices.values()])
 
     def pymbolic_indices(self, indices):
-        idx = tuple()
+        pym_idx = tuple()
         for index in indices:
-            assert index.extent
-            if not index.name:
-                index.name = self.name_gen(self.index_names[index])
-            self.index_extent[index.name] = index.extent
-            idx = idx + (p.Variable(index.name), )
-        return idx
+            name = index.name
+            self.index_extent[name] = index.extent
+            pym_idx += (p.Variable(name), )
+        return pym_idx
+
 
 @contextmanager
-def active_indices(gem_indices, pym_indices, lambda_statement, ctx):
+def active_indices(gem_indices, pym_indices, ctx):
+    """ Context manager function to keep track of active indices.
+    Note: Does not work for recursion, hence cannot be used for statement_for.
+    """
     for gem_index, pym_index in zip(gem_indices, pym_indices):
         ctx.active_indices[gem_index] = pym_index
-    yield lambda_statement(ctx)
+    yield ctx
     for index in gem_indices:
```

Comment:
It's not clear to me why this doesn't work with recursion (you in-place modify `ctx`) so I think:
```python
with active_indices(a, b, ctx):
   do_something(ctx)
   with active_indices(c, d, ctx):
      do_other(ctx)
```

Should see `a -> b` in `do_something` and `a ->b, c -> d` in `do_other`

Reply:
Ok

Reply:
BTW, is it obvious why sometimes one has to do `pym_index = ctx.name_gen(idx)` and sometimes just `pym_index = p.Variable(idx.name)` ?

Reply:
No, this is related to something that Miklos critiqued on the tsslac PR. I name all indices at gem stage so I can just do the latter at the moment. I will change that here as soon as I postponed the index naming to this stage.